test/sanity/code-smell/required-and-default-attributes.py

```diff
@@ -1,19 +1,31 @@
 from __future__ import annotations
 
-import re
 import sys
+import ast
+
+
+class CallVisitor(ast.NodeVisitor):
+    def __init__(self, path):
+        self.path = path
+
+    def visit_Call(self, node):
+        if getattr(node.func, "id", "").endswith("FieldAttribute"):
+            count = 0
+            for kw in node.keywords:
+                if kw.arg in ("default", "required"):
+                    count += 1
+            if count > 1:
+                print(
+                    f"{self.path}:{node.lineno}:{node.col_offset}: use only one of `default` or `required` with `{node.func.id}`"
+                )
```

Comment:
Since we're reworking this test, there are several other things that can be improved as well:

* sorting imports
* including type annotations
* passing the path to `ast.parse` for more helpful error messages
* using `pathlib` to simplify file access
* using `isinstance` instead of attribute checks
* using a comprehension to simplify the logic

With those changes we'd end up with something like this:

```python
from __future__ import annotations

import ast
import pathlib
import sys


class CallVisitor(ast.NodeVisitor):
    def __init__(self, path: str) -> None:
        self.path = path

    def visit_Call(self, node: ast.Call) -> None:
        if isinstance(node.func, ast.Name) and node.func.id.endswith("FieldAttribute"):
            if len([kw for kw in node.keywords if kw.arg in ("default", "required")]) > 1:
                print(f"{self.path}:{node.lineno}:{node.col_offset}: use only one of `default` or `required` with `{node.func.id}`")


def main() -> None:
    for path in sys.argv[1:] or sys.stdin.read().splitlines():
        tree = ast.parse(pathlib.Path(path).read_text(), path)
        CallVisitor(path).visit(tree)


if __name__ == "__main__":
    main()
```

Reply:
Thanks for the suggestion and ready-made code 😉